sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -146,6 +147,56 @@ object EliminateSerialization extends Rule[LogicalPlan] {
 }
 
 /**
+ * Add Filter to left and right of a EquiJoin to filter out rows with null keys.
+ * So we may not need to check nullability of keys while joining. Besides, by filtering
+ * out keys with null, we can also reduce data size in Join.
+ */
+object AddNullFilterForEquiJoin extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case ExtractEquiJoinKeys(Inner, leftKeys, rightKeys, condition, left, right) =>
+      val leftConditions = leftKeys.distinct.map { l =>
+        IsNotNull(l)
+      }.reduceLeft(And)
+
+      val rightConditions = rightKeys.distinct.map { r =>
+        IsNotNull(r)
+      }.reduceLeft(And)
+
+      val keysConditions = ExpressionSet(leftKeys.zip(rightKeys).map { lr =>
+        EqualTo(lr._2, lr._1)
+      }).reduceLeft(And)
+
+      val finalConditions = if (condition.isDefined) {
+        And(keysConditions, condition.get)
+      } else {
+        keysConditions
+      }
+
+      val leftHasAllNotNullPredicate = left.constraints.nonEmpty &&
+        left.constraints.filter(_.isInstanceOf[IsNotNull])
+          .forall(expr => leftConditions.references.intersect(expr.references).nonEmpty)
+
+      val rightHasAllNotNullPredicate = right.constraints.nonEmpty &&
+        right.constraints.filter(_.isInstanceOf[IsNotNull])
+          .forall(expr => rightConditions.references.intersect(expr.references).nonEmpty)
```

Comment:
Because we want to eliminate reading nulls in advance of join operator. If we simply add IsNotNull to the left/right to the conditions, the nulls are still read during performing join.